### ISSUE TITLE
CPR: Classifier Projection Regularization for Continual Learning

### ISSUE THREAD

**Issue by chaerinkong** [reyllama/MIPAL-CL#7]
## TL; DR

- Shoot for wide local minima by applying KL regularization between classifier prediction and uniform distribution (entropy-max).

## Motivation

- Models at wide local minima have better chance of sharing low-loss-region across tasks.

![image](https://user-images.githubusercontent.com/46308407/158291766-a05f4cf1-66ac-4c0f-988a-444d3b024384.png)

## Formulation

- On top of off-the-shelf regularization-based CL method, stack additional regularization that enforces flat local minima.
![image](https://user-images.githubusercontent.com/46308407/158291682-45e5c53f-3462-4c10-b8b2-22a9f81a0364.png)

### Information-geometry Interpretation

- This is equivalent to projecting prediction distribution to a KL-div ball centered at uniform distribution.

![image](https://user-images.githubusercontent.com/46308407/158291826-6dbee7ee-d6db-47fc-b6da-f4aae62ad19f.png)

- By Pythagoras triangular inequality theorem, 
![image](https://user-images.githubusercontent.com/46308407/158292024-58564eb7-ffcd-4d64-b2fc-638e706198d0.png)

- Since KL is non-negative, we see that classifier projection leads to smaller change after each task.
![image](https://user-images.githubusercontent.com/46308407/158292481-f5411408-a24a-4566-8148-3d254b32eeb5.png)

## Evaluation

![image](https://user-images.githubusercontent.com/46308407/158292541-7af0a137-a1be-4b4b-9e3a-a14c0b33a69d.png)
![image](https://user-images.githubusercontent.com/46308407/158292564-2bab1b11-0c0c-4ec9-aa52-369f35c182dd.png)
![image](https://user-images.githubusercontent.com/46308407/158292574-1659c965-820f-412c-8965-813de01071e4.png)

## Comment

- Interesting view on information geometry (first time encountering *probability simplex* to be honest)
